# Creating dataset

In [ ]:
import numpy as np
import pandas as pd
import math
from IPython.display import clear_output
def isprime(num):
    clear_output()
    print("running isprime()")
    for i in range(2,int(num/2)):
        if(num%i==0):
            return 0
    return 1

def find_sum_and_prod_of_digits(n):
    tot=0
    tot1=1
    while(n>0):
        dig = n%10
        tot = tot+dig
        tot1 = tot1*dig
        n=n//10
    return tot,tot1

def prime_dataset_creator(start=None, end=None):
    if(start==None and end==None):
        print("Input Error Found Missing START and END Parameter")
        return None
    elif(start==None):
        print("Input Error Found Missing START Parameter")
        return None
    elif(end==None):
        print("Input Error Found Missing END Parameter")
        return None
    df = pd.DataFrame()
    prime =[]
    num = []
    msb = []
    lsb = []
    log2 = []
    log10 = []
    loge = []
    divbypi = []
    divbye = []
    sqrt = []
    cbrt = []
    sumofdigit = []
    sqrt_sumofdigit = []
    cbrt_sumofdigit = []
    prodofdigit = []
    sqrt_prodofdigit = []
    cbrt_prodofdigit = []
    for i in range(start,end+1):
        #to add the number num
        num.append(i)
        #to add the most significant digit
        msb.append(str(i)[0])
        #to add the least significant digit
        lsb.append(str(i)[-1])
        #to add log2(num)
        log2.append(round(math.log(i,2)))
        #to add log10(num)
        log10.append(round(math.log(i,10)))
        #to add loge(num)
        loge.append(round(math.log(i,math.e)))
        #to add num/pi
        divbypi.append(round(i/math.pi))
        #to add num/e
        divbye.append(round(i/math.e))
        #to add squareroot(num)
        sqrt.append(round(math.sqrt(i)))
        #to add cuberoot(num)
        cbrt.append(round(i**(1./3.)))
        #sod contains sum of digits & pod contains product of digits
        sod,pod = find_sum_and_prod_of_digits(i)
        #sod, squareroot(sod) and cuberoot(sod)
        sumofdigit.append(sod)
        sqrt_sumofdigit.append(round(math.sqrt(sod)))
        cbrt_sumofdigit.append(round(sod**(1./3.)))
        #pod, squareroot(pod) and cuberoot(pod)
        prodofdigit.append(pod)
        sqrt_prodofdigit.append(round(math.sqrt(pod)))
        cbrt_prodofdigit.append(round(pod**(1./3.)))
        #to find if is prime (1 - yes, 0 - no)
        prime.append(isprime(i))
        clear_output()
        print(int((((i-start)/(end-start-1))*100)),"% done - 1 out of 2")
    df['values'] = num
    df['isprime'] = prime
    df['lsb_digit'] = lsb
    df['msb_digit'] = msb
    df['log2'] = log2
    df['log10'] = log10
    df['loge'] = loge
    df['Div_By_Pi'] = divbypi
    df['Div_By_e'] = divbye
    df['sqroot_number'] = sqrt
    df['cuberoot_number'] = cbrt
    df['Sum_Of_Digits'] = sumofdigit
    df['sqrt_Sum_Of_Digits'] = sqrt_sumofdigit
    df['cbrt_Sum_Of_Digits'] = cbrt_sumofdigit
    df['Prod_Of_Digits'] = prodofdigit
    df['sqrt_Prod_Of_Digits'] = sqrt_prodofdigit
    df['cbrt_Prod_Of_Digits'] = cbrt_prodofdigit
    
    # for finding if mod of num by certain numbers and result of num divide by certain numbers
    for j in range(2,20):
        clear_output()
        print(int((((i-start)/(end-start-1))*100)),"% done - 2 out of 2")
        temp = []
        temp1 = []
        for i in range(start,end+1):
            if(i%j==0):
                temp.append(1)
            else:
                temp.append(0)
            temp1.append(round(i/j))# using round for division "div_by" stuff
        name_mod = "Mod_of_"+str(j) 
        df[name_mod]=temp
        name_div = "Div_by_"+str(j) 
        df[name_div]=temp1
    return df 
    
    

In [ ]:
df1 = prime_dataset_creator(start = 2,end = 1000000)
df1.head()

In [ ]:
len(df1)

In [ ]:
%%time
df.to_csv("prime_number_training_dataset.csv", encoding='utf-8', index=False)

# Let's start machine learning

In [1]:
# No warnings
import warnings
warnings.filterwarnings('ignore') # Filter out warnings

# data analysis and wrangling
import pandas as pd
import numpy as np

# visualization
#import matplotlib.pyplot as plt
#%matplotlib inline
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from IPython.display import clear_output
from IPython.display import clear_output

from sklearn.model_selection import train_test_split

# machine learning
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB # Gaussian Naive Bays
from sklearn.linear_model import Perceptron
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import plot_tree

import xgboost as xgb

In [2]:
df = pd.read_csv("prime_number_training_dataset.csv")
df.head()

isprime  values  Mod_of_2  Div_by_2  Mod_of_3  Div_by_3  Mod_of_4  \
0        1       2         1         1         0         0         0   
1        1       3         0         1         1         1         0   
2        0       4         1         2         0         1         1   
3        1       5         0         2         0         1         0   
4        0       6         1         3         1         2         0   

   Div_by_4  Mod_of_5  Div_by_5  ...      loge  productofdigits  \
0         0         0         0  ...  0.693147                2   
1         0         0         0  ...  1.098612                3   
2         1         0         0  ...  1.386294                4   
3         1         1         1  ...  1.609438                5   
4         1         0         1  ...  1.791759                6   

   sqroot_productofdigits  cuberoot_productofdigits  cuberoot_sumofdigits  \
0                1.414214                  1.259921              1.259921   
1                1.732051                  1.442250              1.442250   
2                2.000000                  1.587401              1.587401   
3                2.236068                  1.709976              1.709976   
4                2.449490                  1.817121              1.817121   

   sqroot_sumofdigits  sqroot_number  cuberoot_number   DivByPi   DivBy_e  
0            1.414214       1.414214         1.259921  0.636620  0.735759  
1            1.732051       1.732051         1.442250  0.954930  1.103638  
2            2.000000       2.000000         1.587401  1.273240  1.471518  
3            2.236068       2.236068         1.709976  1.591549  1.839397  
4            2.449490       2.449490         1.817121  1.909859  2.207277  

[5 rows x 53 columns]

In [3]:
df.columns

Index(['isprime', 'values', 'Mod_of_2', 'Div_by_2', 'Mod_of_3', 'Div_by_3',
       'Mod_of_4', 'Div_by_4', 'Mod_of_5', 'Div_by_5', 'Mod_of_6', 'Div_by_6',
       'Mod_of_7', 'Div_by_7', 'Mod_of_8', 'Div_by_8', 'Mod_of_9', 'Div_by_9',
       'Mod_of_10', 'Div_by_10', 'Mod_of_11', 'Div_by_11', 'Mod_of_12',
       'Div_by_12', 'Mod_of_13', 'Div_by_13', 'Mod_of_14', 'Div_by_14',
       'Mod_of_15', 'Div_by_15', 'Mod_of_16', 'Div_by_16', 'Mod_of_17',
       'Div_by_17', 'Mod_of_18', 'Div_by_18', 'Mod_of_19', 'Div_by_19',
       'Sum of digits', 'lsb_digit', 'msb_digit', 'log2', 'log10', 'loge',
       'productofdigits', 'sqroot_productofdigits', 'cuberoot_productofdigits',
       'cuberoot_sumofdigits', 'sqroot_sumofdigits', 'sqroot_number',
       'cuberoot_number', 'DivByPi', 'DivBy_e'],
      dtype='object')

In [4]:
sns.set(style='white', context='notebook', palette='deep')
plt.rcParams[ 'figure.figsize' ] = 20 , 12

pd.set_option('display.max_columns', 100) # Print 100 Pandas columns

# Special distribution plot (will be used later)
def plot_distribution( df , var , target , **kwargs ):
    row = kwargs.get( 'row' , None )
    col = kwargs.get( 'col' , None )
    facet = sns.FacetGrid( df , hue=target , aspect=4 , row = row , col = col )
    facet.map( sns.kdeplot , var , shade= True )
    facet.set( xlim=( 0 , df[ var ].max() ) )
    facet.add_legend()
    plt.tight_layout()

# Initial testing

In [ ]:
np.random.seed(1337) 
X = df.drop("isprime", axis=1) # Training & Validation data
Y = df["isprime"]              # Response / Target Variable
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.20)

In [ ]:
%%time
logreg = LogisticRegression()                                # instantiate
logreg.fit(X_train, Y_train)                                 # fit
Y_pred = logreg.predict(X_test)                               # predict
acc_logreg = sum(Y_pred == Y_test)/len(Y_test)*100                # evaluate

print('Logistic Regression labeling accuracy:', str(round(acc_logreg,2)),'%')

Logistic Regression labeling accuracy: 92.24 %

CPU times: user 7.46 s, sys: 510 ms, total: 7.97 s

Wall time: 8.11 s

In [ ]:
%%time
knn = KNeighborsClassifier(n_neighbors = 3)                  # instantiate
knn.fit(X_train, Y_train)                                    # fit
acc_knn = knn.score(X_test, Y_test)                            # predict + evaluate

print('K-Nearest Neighbors labeling accuracy:', str(round(acc_knn*100,2)),'%')

K-Nearest Neighbors labeling accuracy: 91.01 %

CPU times: user 21.1 s, sys: 350 ms, total: 21.4 s

Wall time: 21.5 s

In [ ]:
%%time
perceptron = Perceptron()                                    # instantiate 
perceptron.fit(X_train, Y_train)                             # fit
acc_perceptron = perceptron.score(X_test, Y_test)                 # predict + evalaute

print('Perceptron labeling accuracy:', str(round(acc_perceptron*100,2)),'%')

Perceptron labeling accuracy: 91.71 %

CPU times: user 2.29 s, sys: 278 ms, total: 2.57 s

Wall time: 2.56 s

In [ ]:
%%time
gradboost = xgb.XGBClassifier(n_estimators=1000)             # instantiate
gradboost.fit(X_train, Y_train)                              # fit
acc_xgboost = gradboost.score(X_test, Y_test)                # predict + evalute

print('XGBoost labeling accuracy:', str(round(acc_xgboost*100,2)),'%')

XGBoost labeling accuracy: 92.4 %
    
CPU times: user 44min 30s, sys: 8.77 s, total: 44min 38s
            
Wall time: 44min 46s

In [ ]:
%%time
random_forest = RandomForestClassifier(n_estimators=500)   # instantiate
random_forest.fit(X_train, Y_train)                         # fit
acc_rf = random_forest.score(X_test, Y_test)                  # predict + evaluate

print('Random Forest labeling accuracy:', str(round(acc_rf*100,2)),'%')

Random Forest labeling accuracy: 91.64 %
    
CPU times: user 34min 48s, sys: 10.6 s, total: 34min 58s
            
Wall time: 35min

# As we got the best results for logistic regression, let's try to improve it further. 
# (We cannot do much with hyperparameters due to binary output)

In [ ]:
%%time
np.random.seed(1337) 
X = df.drop("isprime", axis=1) # Training & Validation data
Y = df["isprime"]           
#X = df.drop(l1, axis=1) # Training & Validation data
Y = df["isprime"]              # Response / Target Variable
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.20)
logreg = LogisticRegression()                                # instantiate
logreg.fit(X_train, Y_train)                                 # fit
Y_pred = logreg.predict(X_test)                               # predict
acc_logreg = sum(Y_pred == Y_test)/len(Y_test)*100                # evaluate
print(acc_logreg)

92.23599999999999

CPU times: user 7.44 s, sys: 600 ms, total: 8.04 s

Wall time: 8.01 s

In [ ]:
%%time
np.random.seed(1337) # set random seed for reproducibility
lq = [0.15,0.18,0.20]
num1list = [15,20,50]
for num1 in num1list:
    finalans = []
    for j in lq:    
        m=[]
        for i in range(0,num1):    
            X = df.drop("isprime", axis=1) # Training & Validation data
            Y = df["isprime"]           
            #X = df.drop(l1, axis=1) # Training & Validation data
            Y = df["isprime"]              # Response / Target Variable
            X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=j)
            logreg = LogisticRegression()                                # instantiate
            logreg.fit(X_train, Y_train)                                 # fit
            Y_pred = logreg.predict(X_test)                               # predict
            acc_logreg = sum(Y_pred == Y_test)/len(Y_test)*100                # evaluate
            m.append(acc_logreg)
            #clear_output()
            #print(j,"->",int(i*5),"%")
        finalans.append(sum(m)/num1)
    print(num1,"->",finalans)

# Results

                   15%             18%            20%   
TRY 3 -> 50 -> [92.15209333333335, 92.16923333333335, 92.14147999999999]

TRY 2 -> 20 -> [92.14289999999998, 92.15105555555556, 92.15309999999998]

TRY 1 -> 15 -> [92.13986666666668, 92.15485185185184, 92.16933333333333]

CPU times: user 33min 57s, sys: 3min 13s, total: 37min 11s
Wall time: 37min 12s

We will take .2 as test size because of above results

# Attempt to check which column is useless in determining prime number

## How will we know?

We know that the accuracy for original dataset with all the columns is 92.16933333333333%.

If the score remains same or increases when a column is dropped means it was useless.

We will remove each column, run for 50 times and find accuracy mean and store it.

In [ ]:
np.random.seed(1337) 
namelist=[]
accuracylist=[]
#X = df.drop("isprime", axis=1) # Training & Validation data
Y = df["isprime"]              # Response / Target Variable
for i in range(1,53):
    macc=[]
    print("(",i,")->",df.columns[i],end=" -> ")
    namelist.append(df.columns[i])
    for k in range(0,20):
        X = df.drop("isprime", axis=1)
        Xtemp = X.drop(df.columns[i],axis=1)
        X_train, X_test, Y_train, Y_test = train_test_split(Xtemp, Y, test_size=0.20)
        logreg = LogisticRegression()                                # instantiate
        logreg.fit(X_train, Y_train)                                 # fit
        Y_pred = logreg.predict(X_test)                               # predict
        acc_logreg = sum(Y_pred == Y_test)/len(Y_test)*100# evaluate
        macc.append(acc_logreg)
    io = sum(macc)/20
    accuracylist.append(io)
    print('Accuracy:', str(io),'%\n')

( 1 )-> values -> Accuracy: 92.13759999999999 %

( 2 )-> Mod_of_2 -> Accuracy: 92.156375 %

( 3 )-> Div_by_2 -> Accuracy: 92.16035000000001 %

( 4 )-> Mod_of_3 -> Accuracy: 92.1446 %

( 5 )-> Div_by_3 -> Accuracy: 92.15859999999999 %

( 6 )-> Mod_of_4 -> Accuracy: 92.1457 %

( 7 )-> Div_by_4 -> Accuracy: 92.137 %

( 8 )-> Mod_of_5 -> Accuracy: 92.1682 %

( 9 )-> Div_by_5 -> Accuracy: 92.16452500000001 %

( 10 )-> Mod_of_6 -> Accuracy: 92.174575 %

( 11 )-> Div_by_6 -> Accuracy: 92.15292500000001 %

( 12 )-> Mod_of_7 -> Accuracy: 92.128725 %

( 13 )-> Div_by_7 -> Accuracy: 92.14565000000002 %

( 14 )-> Mod_of_8 -> Accuracy: 92.12269999999998 %

( 15 )-> Div_by_8 -> Accuracy: 92.15655000000001 %

( 16 )-> Mod_of_9 -> Accuracy: 92.15082500000001 %

( 17 )-> Div_by_9 -> Accuracy: 92.14365000000001 %

( 18 )-> Mod_of_10 -> Accuracy: 92.1531 %

( 19 )-> Div_by_10 -> Accuracy: 92.14815000000002 %

( 20 )-> Mod_of_11 -> Accuracy: 92.1478 %

( 21 )-> Div_by_11 -> Accuracy: 92.15795 %

( 22 )-> 

( 1 )-> values -> Accuracy: 92.12986666666666 %

( 2 )-> Mod_of_2 -> Accuracy: 92.15450000000003 %

( 3 )-> Div_by_2 -> Accuracy: 92.16933333333333 %

( 4 )-> Mod_of_3 -> Accuracy: 92.15206666666668 %

( 5 )-> Div_by_3 -> Accuracy: 92.14566666666668 %

( 6 )-> Mod_of_4 -> Accuracy: 92.15326666666665 %

( 7 )-> Div_by_4 -> Accuracy: 92.15210000000002 %

( 8 )-> Mod_of_5 -> Accuracy: 92.14750000000001 %

( 9 )-> Div_by_5 -> Accuracy: 92.1385 %

( 10 )-> Mod_of_6 -> Accuracy: 92.16276666666666 %

( 11 )-> Div_by_6 -> Accuracy: 92.15329999999999 %

( 12 )-> Mod_of_7 -> Accuracy: 92.17173333333334 %

( 13 )-> Div_by_7 -> Accuracy: 92.17676666666667 %

( 14 )-> Mod_of_8 -> Accuracy: 92.15693333333333 %

( 15 )-> Div_by_8 -> Accuracy: 92.13910000000001 %

( 16 )-> Mod_of_9 -> Accuracy: 92.13550000000002 %

( 17 )-> Div_by_9 -> Accuracy: 92.14923333333334 %

( 18 )-> Mod_of_10 -> Accuracy: 92.12596666666667 %

( 19 )-> Div_by_10 -> Accuracy: 92.13049999999998 %

( 20 )-> Mod_of_11 -> Accuracy: 92.16083333333331 %

( 21 )-> Div_by_11 -> Accuracy: 92.13776666666668 %

( 22 )-> Mod_of_12 -> Accuracy: 92.15536666666667 %

( 23 )-> Div_by_12 -> Accuracy: 92.13496666666667 %

( 24 )-> Mod_of_13 -> Accuracy: 92.16866666666667 %

( 25 )-> Div_by_13 -> Accuracy: 92.14306666666668 %

( 26 )-> Mod_of_14 -> Accuracy: 92.14063333333334 %

( 27 )-> Div_by_14 -> Accuracy: 92.15673333333332 %

( 28 )-> Mod_of_15 -> Accuracy: 92.16476666666668 %

( 29 )-> Div_by_15 -> Accuracy: 92.17233333333333 %

( 30 )-> Mod_of_16 -> Accuracy: 92.17246666666666 %

( 31 )-> Div_by_16 -> Accuracy: 92.1604 %

( 32 )-> Mod_of_17 -> Accuracy: 92.16293333333333 %

( 33 )-> Div_by_17 -> Accuracy: 92.14563333333334 %

( 34 )-> Mod_of_18 -> Accuracy: 92.14373333333332 %

( 35 )-> Div_by_18 -> Accuracy: 92.14670000000001 %

( 36 )-> Mod_of_19 -> Accuracy: 92.15450000000001 %

( 37 )-> Div_by_19 -> Accuracy: 92.14723333333336 %

( 38 )-> Sum of digits -> Accuracy: 92.17563333333334 %

( 39 )-> lsb_digit -> Accuracy: 92.1278 %

( 40 )-> msb_digit -> Accuracy: 92.13633333333334 %

( 41 )-> log2 -> Accuracy: 92.14859999999999 %

( 42 )-> log10 -> Accuracy: 92.15209999999999 %

( 43 )-> loge -> Accuracy: 92.1237 %

( 44 )-> productofdigits -> Accuracy: 92.14073333333333 %

( 45 )-> sqroot_productofdigits -> Accuracy: 92.16586666666667 %

( 46 )-> cuberoot_productofdigits -> Accuracy: 92.15979999999999 %

( 47 )-> cuberoot_sumofdigits -> Accuracy: 92.13916666666667 %

( 48 )-> sqroot_sumofdigits -> Accuracy: 92.13156666666667 %

( 49 )-> sqroot_number -> Accuracy: 92.13626666666669 %

( 50 )-> cuberoot_number -> Accuracy: 92.15683333333332 %

( 51 )-> DivByPi -> Accuracy: 92.14483333333334 %

( 52 )-> DivBy_e -> Accuracy: 92.14696666666666 %


In [ ]:
namelist

In [ ]:
for i in range(0,len(accuracylist)-1):
    for j in range(i+1,len(accuracylist)):
        if(accuracylist[i]>accuracylist[j]):
            temp = accuracylist[i]
            accuracylist[i] = accuracylist[j]
            accuracylist[j] = temp
            temp = namelist[i]
            namelist[i] = namelist[j]
            namelist[j] = temp
line = []
for i in range(0,len(namelist)):
    line.append(92.16933333333333)
plt.plot(line,namelist,linestyle='--')
plt.plot(accuracylist,namelist)
plt.scatter(accuracylist,namelist)
plt.show()  # Note: we don't hav

## Image is there in home -> name -> "logistic graph"

In [ ]:
namelist2 = []
accuracylist2 = []
for i in range(0,len(namelist)):
    if accuracylist[i] >= 92.16933333333333:
        accuracylist2.append(accuracylist[i])
        namelist2.append(namelist[i])
print(namelist2)
print(accuracylist2)

['Div_by_2', 'Mod_of_7', 'Div_by_15', 'Mod_of_16', 'Sum of digits', 'Div_by_7']

[92.16933333333333, 92.17173333333334, 92.17233333333333, 92.17246666666666, 92.17563333333334, 92.17676666666667]

# Confirming the hypothesis - 

list l1 columns should be removed.

In [ ]:
namelist2.append("isprime")
namelist2

['Div_by_2',
 'Mod_of_7',
 'Div_by_15',
 'Mod_of_16',
 'Sum of digits',
 'Div_by_7',
 'isprime']

In [ ]:
%%time
np.random.seed(1337) # set random seed for reproducibility
m1=[]
numm = 15
for i in range(0,numm):    
    X = df.drop(namelist2, axis=1) # Training & Validation data
    Y = df["isprime"]          
    #X = df.drop(l1, axis=1) # Training & Validation data
    Y = df["isprime"]              # Response / Target Variable
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.20)
    logreg = LogisticRegression()                                # instantiate
    logreg.fit(X_train, Y_train)                                 # fit
    Y_pred = logreg.predict(X_test)                               # predict
    acc_logreg = sum(Y_pred == Y_test)/len(Y_test)*100                # evaluate
    m1.append(acc_logreg)
    clear_output()
    print(int(i+1)," out of ", numm)

15  out of  15

CPU times: user 1min 48s, sys: 9.51 s, total: 1min 57s
            
Wall time: 1min 58s

In [ ]:
ans = sum(m1)/numm
print(ans)

92.12986666666666

#### Result for logistic regression is not that promising due to biasing present in dataset. The dataset contains many values for non prime numbers and few values of prime numbers.


# Now, We will try to reduce the bias and re-run the same code, still of results are bad then we will discard this model and move to more sophesticated one like decision tree or random forest.

# -----------------------------------------------------------------------
# for mudit -> Try generalization approach, have approximately equal percentage of prime and non prime numbers.

In [ ]:
ctr=0
for i in range(0,len(df)):
    if(df['isprime'][i]==1):
        ctr=ctr+1
print("primes -",ctr)
print("not primes -",len(df)-ctr)

In [ ]:
78498/921500

In [ ]:
target_count = df['isprime'].value_counts()
target_count

In [ ]:

# What is base line for prediction accuracy?

target_count[0]/(sum(target_count))